eventcore/src/executor.rs

```diff
@@ -674,147 +716,257 @@ where
         for<'a> <C::Event as TryFrom<&'a ES::Event>>::Error: std::fmt::Display,
         ES::Event: From<C::Event> + Clone + serde::Serialize,
     {
-        let mut stream_ids = command.read_streams();
+        let mut context = StreamDiscoveryContext::new(
+            command.read_streams(),
+            options.max_stream_discovery_iterations,
+        );
         let mut stream_resolver = StreamResolver::new();
-        let mut iteration = 0;
-        let max_iterations = options.max_stream_discovery_iterations;
 
         loop {
```

Comment:
Not sure if this is something that should be addressed in this PR or not, but I'd love to find a way to get rid of this procedural loop in favor of a more functional style of iteration.

Reply:
It might make sense to do now. There are a lot of downstream refactorings from here that might be rendered completely redundant if we go this route.